spec-abstract-noclosure/MC_ffg.tla

```diff
@@ -0,0 +1,101 @@
+----------------------------- MODULE MC_ffg -----------------------------
+
+EXTENDS FiniteSets
+
+\* @type: Int;
+MAX_BLOCK_SLOT == 5
+\* @type: Int;
+MAX_BLOCK_BODY == 3
+
+\* @type: Set(Str);
+VALIDATORS == {"V1", "V2", "V3", "V4"}
+
+N == 4
+
+VARIABLES
+    \* the set of all_blocks
+    \* @type: Set($block);
+    all_blocks,
+    \* the set of blocks on chain 1
+    \* @type: Set($block);
+    chain1,
+    \* the latest block on chain 1
+    \* @type: $block;
+    chain1_tip,
+    \* the set of blocks on chain 2
+    \* @type: Set($block);
+    chain2,
+    \* the latest block on chain 2
+    \* @type: $block;
+    chain2_tip,
+    \* If chain2_fork_block_number is not equal to 0,
+    \* then chain2 is a fork of chain1 starting at chain2_fork_block_number
+    \* @type: Int;
+    chain2_fork_block_number,
+    \* @type: Set($ffgVote);
+    ffg_votes,
+    \* @type: Set($vote);
+    votes,
+    \* @type: Set($checkpoint);
+    justified_checkpoints
+
+
+INSTANCE ffg
+
+Abs(x) == IF x >= 0 THEN x ELSE -x
+
+IndInv ==
+    /\ -MAX_BLOCK_BODY <= chain2_fork_block_number /\ chain2_fork_block_number <= 0
+    /\ all_blocks = chain1 \union chain2
+    \* chain1_tip is the maximum block in chain 1
+    /\ \A b \in chain1: b.body <= chain1_tip.body
+    \* block numbers on chain 1 simply go from 0 to chain1_tip.body
+    /\ \A b1, b2 \in chain1:
+        /\ b1.body >= 0 /\ b1.body <= chain1_tip.body
+        /\ (b1.body >= b2.body) <=> (b1.slot >= b2.slot)
+    \* there are no gaps in the block numbers
+    /\ \A i \in 0..MAX_BLOCK_BODY:
+        i <= chain1_tip.body => \E b \in chain1: b.body = i
+    \* chain2_tip is the maximum block in chain 2
```

Comment:
Is chain2_tip the head of chain2

Reply:
It's the latest block in `chain2`, that is, the block with the largest slot.